Directly download commercial ip geolocation databases from providers

joegallo opened the pull request:
Adds a feature for downloading commercial ip geolocation databases directly from different providers and exposing those databases to the `geoip` processor -- at present the only supported provider is [MaxMind](https://www.maxmind.com/).

joegallo committed:
A small tidiness refactor of the GeoIpTaskState's Metadata (#110553)

masseyke committed:
Adding a unit test for GeoIpDownloader.cleanDatabases (#110650) (#110675)

Co-authored-by: Joe Gallo <joegallo@gmail.com>

masseyke committed:
Removing the use of Stream::peek from GeoIpDownloader::cleanDatabases (#110666) (#110679)

joegallo committed:
WIP squashed history of dev work so far

joegallo committed:
Rename this class

joegallo committed:
Tweak these constants to better shape the state

Some cluster state customs have singular nouns and some have plural,
and I think plural is more natural for this case.

joegallo committed:
Make it possible to override the endpoint for tests

joegallo committed:
Log and skip if there are no credentials

masseyke committed:
Updating EnterpriseGeoIpDownloaderLicenseListener to not start task if no database metadata is present

joegallo committed:
Schedule the next run at the very beginning of the run

to work around a race condition on the first run (when there is no
next run already scheduled).

joegallo committed:
Guard these contains operations for the no-change case

joegallo committed:
Remove the conflict between the two geoip downloaders

masseyke committed:
Adding EnterpriseGeoIpTaskStateSerializationTests

joegallo committed:
Rename this setting and the associated constant

From ingest.geoip.downloader.maxmind.default.license_key to just
ingest.geoip.downloader.maxmind.license_key.

joegallo committed:
Address todos and comments, and tidy up logging

joegallo committed:
Bump to Priority.NORMAL and passed hallway review

masseyke committed:
Merge branch 'enterprise-downloader' of github.com:joegallo/elasticsearch into enterprise-downloader

masseyke committed:
Adding IngestGeoIpMetadataTests (and fixing a bug it found)

joegallo committed:
This reads slightly better in terms of intent

joegallo committed:
Switch from a setting to a field for the account_id

In this version Maxmind is very much a hardcoded configuration -- I
think we might want a NamedWriteable here rather than merely a
Writeable, but this allows us to keep making progress.

masseyke committed:
Adding yaml rest tests

joegallo committed:
Adjust the new NodeFeature slightly

joegallo committed:
Collapse the two versions and add an alias

masseyke committed:
beginnings of a unit test for EnterpriseGeoIpDownloaderTests

joegallo committed:
Override the default maxmind endpoint for the yaml tests

joegallo committed:
Merge remote-tracking branch 'masseyke/enterprise-downloader' into enterprise-downloader

joegallo committed:
Merge branch 'main' into enterprise-downloader

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Update docs/changelog/110844.yaml

joegallo committed:
These are non-operator actions

masseyke committed:
Moving NodeFeature and FeatureSpecification to server to avoid build errors

masseyke committed:
adding EnterpriseGeoIpDownloaderIT

joegallo committed:
Enhance this test

This is not necessary for this PR, but we had a failure here and I've
always hated the failure messages from this class because it makes me
compare the lists of files myself and do the mental accounting of
what's right and what's wrong, etc. Anyway, using `assertThat` gives
us a nicer error message about exactly what was wrong with the
returned key set.

masseyke committed:
cleaning up EnterpriseGeoIpDownloaderIT

masseyke committed:
Merge branch 'enterprise-downloader' of github.com:joegallo/elasticsearch into enterprise-downloader

joegallo committed:
Fix a typo

joegallo committed:
Merge branch 'main' into enterprise-downloader

masseyke committed:
creating a new EnterpriseGeoIpHttpFixture

masseyke committed:
merging

masseyke committed:
removing enterprise logic from standard geo fixture

masseyke committed:
removing whitespace change

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportPutDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,198 @@
+        }
+    }
+
+    private static void validatePrerequisites(DatabaseConfiguration database, ClusterState state) {
```
Can you add a unit test for this? (I haven't double-checked that you didn't already, but since it's marked `private` I assume there isn't one)

dakrone commented on modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/GeoIpDownloaderTests.java:
```diff
@@ -426,6 +426,55 @@ void deleteOldChunks(String name, int firstChunk) {
+        GeoIpDownloaderStats stats = geoIpDownloader.getStatus();
+        assertEquals(1, stats.getExpiredDatabases());
+        assertEquals(2, count.get()); // somewhat surprising, not necessarily wrong
+        assertEquals(18, geoIpDownloader.state.getDatabases().get("test.mmdb").lastCheck()); // highly surprising, seems wrong
```
Can you explain why this is surprising and seems wrong? Does it need more investigation?

masseyke replied:
@joegallo  and I *think* that this line was probably a mistake: https://github.com/elastic/elasticsearch/blob/main/modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDownloader.java#L330. We can't think of a valid reason to subtract 1 from the last checked millis every time we run `cleanDatabases`. We suspect that the intention was to set it to `-1` as a marker for a future check. Right now it doesn't really do any harm, and it was outside of the scope of this PR. So we just tested the behavior as-is, with a note that we might want to come back to it at some point. There had been no code coverage on this behavior before that I could fine.

dakrone commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseDownloaderPlugin.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Collection;
+import java.util.List;
+
+public class EnterpriseDownloaderPlugin extends Plugin {
```
A small description of what this plugin does in the javadoc could be helpful here.

dakrone commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    void listenForLicenseStateChanges() {
+        assert licenseStateListenerRegistered == false;
```
Can you add an assert message? (nothing worse than an empty failure way later on in life)

dakrone commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java:
```diff
@@ -0,0 +1,154 @@
+
+    @Override
+    public void licenseStateChanged() {
+        assert licenseStateListenerRegistered;
```
Same here about an assert message

masseyke replied:
I'm going to just delete this assert because it doesn't seem valuable (it asserts that init() has been called before `licenseStateChanged` is called... but init is what registers this thing as a license state listener. So it doesn't seem like a problem to worry about.

dakrone commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java:
```diff
@@ -0,0 +1,154 @@
+
+    void listenForLicenseStateChanges() {
+        assert licenseStateListenerRegistered == false;
+        licenseState.addListener(this);
+        licenseStateListenerRegistered = true;
```
I think we might want to swap the order of these two, otherwise we could potentially hit the assert in `licenseStateChanged`. E.g., if the listener were added, then the were to pause before setting the value to `true`, and then `licenseStateChanged` got called from another thread prior to this value being updated.

What do you think?

masseyke replied:
Now that it's only used to make sure we don't call init() twice (see my comment about the related assertion below), I don't think it matters. I'll move it anyway though.

dakrone commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java:
```diff
@@ -0,0 +1,154 @@
+        ) == false;
+        /*
+         * We don't want to potentially start the task on every cluster state change, so only maybeUpdateTaskState if this cluster change
+         * event involved the addition of custom geoip metadata OR a master node change
```
```suggestion
         * event involved the modification of custom geoip metadata OR a master node change
```

joegallo replied:
Reword accepted, committed as 8145a58957b

dakrone commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java:
```diff
@@ -0,0 +1,154 @@
+        return state.metadata().custom(INGEST_GEOIP_CUSTOM_METADATA_TYPE) != null;
+    }
+
+    private boolean onMasterNode(ClusterState state) {
+        // We should only start/stop task from single node, master is the best as it will go through it anyway
+        return state.nodes().isLocalNodeElectedMaster();
+    }
```
This method is only called from one place, and is only a single line, so we could probably remove the method and just use `state.nodes().isLocalNodeElectedMaster()` directly?

dakrone commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
```
Why do we need to listen for cluster changed events in addition to listening to the license state changes? Is there a reason that listening only to license state changes isn't enough?

masseyke replied:
Our goal was to keep the impact of this change minimal for users who don't use enterprise geoip databases. So we don't start the task until we see that (1) the license is valid and (2) at least one enterprise geoip database has been configured (via RestPutDatabaseConfigurationAction).
Do you think we ought to start the task regardless of whether a database is configured?

dakrone replied:
Ahh okay, that makes sense, I could see it either direction (either the simplicity of always starting, or the lightweightness of not starting it unless necessary). I think this is fine as-is for now then.

dakrone commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java:
```diff
@@ -0,0 +1,154 @@
+        persistentTasksService.sendRemoveRequest(ENTERPRISE_GEOIP_DOWNLOADER, MASTER_TIMEOUT, listener);
+    }
+
+    private boolean hasIngestGeoIpMetadata(ClusterState state) {
+        return state.metadata().custom(INGEST_GEOIP_CUSTOM_METADATA_TYPE) != null;
+    }
```
This is another single-use single-line helper, I think it's clearer to collapse it so that a reader doesn't have to jump to a different place to see the implementation?

dakrone commented on modules/ingest-geoip/src/internalClusterTest/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloaderIT.java:
```diff
@@ -0,0 +1,187 @@
+
+    @SuppressWarnings("unchecked")
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return CollectionUtils.appendToCopyNoNullElements(super.nodePlugins(), IngestGeoIpPlugin.class, ReindexPlugin.class);
```
What do we use `ReindexPlugin` for? I don't see it used anywhere in this file?

joegallo replied:
I added a comment in 97afb241c2b -- interestingly enough, `ReindexPlugin` provides the ability to `_delete_by_query`, and `modules/ingest-geoip` uses that.

masseyke replied:
GeoIp databases are deleted using the delete-by-query action. That action is in server, but it's registered in the ReindexPlugin. So the code blows up if that plugin is not available.

dakrone replied:
Ahh interesting, thanks for the background.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseNodeService.java:
```diff
@@ -238,7 +240,7 @@ public void close() throws IOException {
     }
 
-    void checkDatabases(ClusterState state) {
+    void checkDatabases(ClusterState state) { // cluster state listener
```
What does the comment mean? Can you elaborate?

joegallo replied:
Eh, it was just a reminder from myself to myself, I just removed it in a5906b3d63d.

dakrone reviewed (changes requested):
I'm not done with this (41/50 files viewed), but I wanted to submit an in-progress review for async work.

joegallo committed:
Drop unnecessary scope

joegallo committed:
Add a comment about why the ReindexPlugin is necessary

joegallo committed:
Remove a comment

joegallo committed:
Rename this variable

joegallo committed:
Drop this unnecessary class

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/PutDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,93 @@
+            Request create(DatabaseConfiguration database);
+
+            String getId();
+        }
```
Yeah, removed in daa77591b95 -- it was necessary in a previous version of this code, but it's not anymore.

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/PutDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,93 @@
+
+        private final DatabaseConfiguration database;
+
+        public Request(TimeValue masterNodeTimeout, TimeValue ackTimeout, DatabaseConfiguration lifecycle) {
```
Yarp, fixed in 314bd250da1.

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/IngestGeoIpMetadata.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    public IngestGeoIpMetadata(StreamInput in) throws IOException {
+        {
```
Correct, fixed in 0110daccc02.

joegallo committed:
Reword this comment

masseyke committed:
small changes to EnterpriseDownloaderPlugin and EnterpriseGeoIpDownloaderLicenseListener for code review feedback

masseyke committed:
fixing javadoc link

masseyke committed:
undoing accidental revert of commeent change

joegallo committed:
Allow unknown fields on this parser

joegallo committed:
Add a javadoc for PasswordAuthenticationHolder

joegallo committed:
Remove unused method

joegallo committed:
Simplify serialization and deserialization

joegallo committed:
Rewrite the get logic slightly

joegallo committed:
Use the correct transport version here

joegallo committed:
Handle deletes slightly differently

masseyke committed:
Improving and testing database configuration validation

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        // this is for injecting the state in GeoIpDownloaderTaskExecutor#nodeOperation just after the task instance has been created
+        // by the PersistentTasksNodeService -- since the GeoIpDownloader is newly created, the state will be null, and the passed-in
+        // state cannot be null
+        assert this.state == null;
+        assert state != null;
```
Can you add messages to these asserts? Should the second one be an actual check?

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        var clusterState = clusterService.state();
+        var geoipIndex = clusterState.getMetadata().getIndicesLookup().get(EnterpriseGeoIpDownloader.DATABASES_INDEX);
+        if (geoipIndex != null) {
+            logger.trace("The {} index is not null", EnterpriseGeoIpDownloader.DATABASES_INDEX);
```
This looks a big like dev logging to me, perhaps we should make it something more like `"the enterprise geoip index [{}] exists"`, what do you think?

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        if (geoipIndex != null) {
+            logger.trace("The {} index is not null", EnterpriseGeoIpDownloader.DATABASES_INDEX);
+            if (clusterState.getRoutingTable().index(geoipIndex.getWriteIndex()).allPrimaryShardsActive() == false) {
+                throw new ElasticsearchException("not all primary shards of [" + DATABASES_INDEX + "] index are active");
```
Should this be an actual exception (with the error logging this would cause), or should it be more of a warning/ignorable thing? I'm worried about spitting a bunch of errors and logging whenever a node restarts.

masseyke replied:
This was copy/pasted from GeoIpDownloader (this change added in https://github.com/elastic/elasticsearch/pull/86842). We were attempting to keep them pretty similar so that they could be merged back into one thing in the future. This seems like a small enough difference though that I could just change it to a debug level message.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        // else, just download everything
+        boolean addedSomething = false;
+        {
+            EnterpriseGeoIpTaskState _state = state;
+            Set<String> metas = Set.copyOf(_state.getDatabases().keySet());
```
This seems odd to me to copy this only to use it once? Is there a reason you did this? I don't think we have it for consistency since you only used it once in this block?

masseyke replied:
I'm going to leave this one for @joegallo to comment on. Maybe he's avoiding some race condition since this method is running in the cluster state applier thread, and something else could be modifying the task state at the same time?

joegallo replied:
I'll drop it -- the underlying map is immutable anyway, so it's not like we need to worry about the keyset mutating out from underneath us. I imagine it was just me thinking defensively for a little bit there.

joegallo replied:
Dropped in 39416f0d751.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+                final String id = entry.getKey();
+                DatabaseConfiguration database = entry.getValue().database();
+                if (metas.contains(database.name() + ".mmdb") == false) {
+                    logger.debug("A new database appeared [{}]", database.name());
```
:pokemon: A wild database appeared!

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+                        }
+                    }
+
+                    addedSomething = true;
```
I think this can be moved into the body of the `else { … }` clause above. Currently this treats skipping the download because credentials are missing as adding something, but we may not actually be adding something.

masseyke replied:
I'm going to leave this to @joegallo in case I don't fully understand the intention here.

joegallo replied:
Fair enough, I changed this in 404bb018c60.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+                .map(c -> c.database().name() + ".mmdb")
+                .collect(Collectors.toSet());
+            EnterpriseGeoIpTaskState _state = state;
+            Collection<Map.Entry<String, Metadata>> metas = List.copyOf(_state.getDatabases().entrySet());
```
I left a comment about unexpected `Map.Entry` behavior before, should this one also just be converted into a collection of Tuples?

masseyke replied:
Yeah it probably makes sense here since we're making a copy of it.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+                }
+            }
+            state = _state;
+            updateTaskState();
```
Should this unconditionally update the state, or should this be inside a `if (droppedSomething) { updateTaskState(); }` block?

masseyke replied:
I think you are right (and I committed that change). @joegallo can chime in if there's something to the logic we're missing.

joegallo replied:
Nah, I think I'd intended it to be what you ended up doing, I just didn't see the miss. Thanks for catching it!

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        }
+    }
+
+    void processDatabase(PasswordAuthentication auth, String id, DatabaseConfiguration database) throws IOException {
```
Can you add javadoc for what "processing" a database means?

dakrone replied:
Also, you passed in both the ID and the database, but the `DatabaseConfiguration` already has an id, should that one be used instead?

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        final String sha256Url = downloadUrl(name, "tar.gz.sha256");
+        final String tgzUrl = downloadUrl(name, "tar.gz");
+
+        final Pattern checksumPattern = Pattern.compile("(\\w{64})\\s\\s(.*)");
```
This can be a static final variable at the class level instead of compiling it every time I think

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        processDatabase(auth, name + ".mmdb", sha256, tgzUrl);
+    }
+
+    private void processDatabase(PasswordAuthentication auth, String name, String sha256, String url) {
```
Same here about javadoc, since the two `processDatabase` calls do different things (it may be useful to name them different things)

masseyke replied:
I added javadocs. I'll leave it to @joegallo if he wants to come up with meaningful distinct names for these two methods.

joegallo replied:
I'd prefer we didn't rename them -- there's a single implementation peeking out from underneath both of them that didn't get DRYed out on this PR. It is my intention to do that DRY-ing on a subsequent PR. I'll give them unique names for this PR if you insist, but they'll still be DRYed into a single name (and implementation) within month from now anyway.

dakrone replied:
Leaving it is fine with me, can you add javadoc to both (even though they're private), so someone can tell how they differ at a glance?

joegallo replied:
Certainly!

joegallo replied:
Keith added javadocs in 225513ab6d6 and I expanded them slightly in 3f3a7c0849e.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        client.execute(
+            DeleteByQueryAction.INSTANCE,
+            request,
+            ActionListener.wrap(r -> {}, e -> logger.warn("could not delete old chunks for geoip database [" + name + "]", e))
```
If this deletion fails, we'll just retry it the next time around, right?

joegallo replied:
Yes, that's something I really like about the existing logic in `GeoIpDownloader` -- because it deletes all old chunks for the name (every chunk less than the current `firstChunk`), it'll just retry on a later update of the same database if one of these should happen to fail. It's clever!

For that matter, the way it's implemented doesn't rely on chunks being deleted from the database ever -- it's just a GC process, besides eating up disk space there's no harm in keeping old chunks around forever (so to speak).

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+    }
+
+    // visible for testing
+    Tuple<Integer, String> indexChunks(
```
It may eventually be nice to re-use all the pieces from `GeoIpDownloader` rather than duplicate them. (Not now though, but at a later time). Maybe we could add that as a comment?

masseyke replied:
Yeah I think the intention is to share a lot more between them (and possibly have them be the same thing). But given tight timelines we didn't want to accidentally break the widely-used GeoIpDownloader in this release.

joegallo replied:
Yes, exactly this.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        // there's a race condition between here and requestReschedule. originally this scheduleNextRun call was at the end of this
+        // block, but remember that updateDatabases can take seconds to run (it's downloading bytes from the internet), and so during the
+        // very first run there would be no future run scheduled to reschedule in requestReschedule. which meant that if you went from zero
+        // to N(>=2) databases in quick succession, then the all but the first database wouldn't necessarily get downloaded, because the
```
Random typo (I think):

```suggestion
        // to N(>=2) databases in quick succession, then all but the first database wouldn't necessarily get downloaded, because the
```

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+    }
+
+    private void cleanDatabases() {
+        // this cleanDatabases logic is wrong, it'll deleteOldChunks() repeatedly forever,
+        // and it counts down the lastCheck a millisecond at a time which seems like nonsense to me.
```
Uh? Should we be adding this if it's the wrong logic? Or is this comment out of date?

joegallo replied:
I removed the comments and opened a ticket for it. It's wrong on the donor side, and this is not-yet-DRYed code. https://github.com/elastic/elasticsearch/issues/110954

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+        List<Map.Entry<String, Metadata>> expiredDatabases = state.getDatabases()
+            .entrySet()
+            .stream()
+            .filter(e -> e.getValue().isValid(clusterService.state().metadata().settings()) == false)
```
Can we rename the `isValid` method to something like `isNewEnough` so that it's clear from reading its use what it's checking? I had to go look it up while reading this to see what it was doing.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java:
```diff
@@ -0,0 +1,449 @@
+            String name = e.getKey();
+            Metadata meta = e.getValue();
+            deleteOldChunks(name, meta.lastChunk() + 1);
+            // the lastCheck() - 1 bit here is a WAT
```
Can you explain in this comment why it's necessary?

joegallo replied:
Same as previous -- I dropped the comments and opened https://github.com/elastic/elasticsearch/issues/110954.

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloaderTaskExecutor.java:
```diff
@@ -0,0 +1,248 @@
+        }
+    }
+
+    /**
+     * Extracts the {@link SecureSettings}` out of the passed in {@link Settings} object. The {@code Setting} argument has to have the
+     * {@code SecureSettings} open/available. Normally {@code SecureSettings} are available only under specific callstacks (eg. during node
+     * initialization or during a `reload` call). The returned copy can be reused freely as it will never be closed (this is a bit of
+     * cheating, but it is necessary in this specific circumstance). Only works for secure settings of type string (not file).
+     *
+     * @param source               A {@code Settings} object with its {@code SecureSettings} open/available.
+     * @param securePluginSettings The list of settings to copy.
+     * @return A copy of the {@code SecureSettings} of the passed in {@code Settings} argument.
+     */
+    private static SecureSettings extractSecureSettings(Settings source, List<Setting<?>> securePluginSettings)
+        throws GeneralSecurityException {
+        // get the secure settings out
+        final SecureSettings sourceSecureSettings = Settings.builder().put(source, true).getSecureSettings();
+        // filter and cache them...
+        final Map<String, Tuple<SecureString, byte[]>> cache = new HashMap<>();
+        if (sourceSecureSettings != null && securePluginSettings != null) {
+            for (final String settingKey : sourceSecureSettings.getSettingNames()) {
+                for (final Setting<?> secureSetting : securePluginSettings) {
+                    if (secureSetting.match(settingKey)) {
+                        cache.put(
+                            settingKey,
+                            new Tuple<>(sourceSecureSettings.getString(settingKey), sourceSecureSettings.getSHA256Digest(settingKey))
+                        );
+                    }
+                }
+            }
+        }
```
This part I'm not familiar at all with, maybe someone like @jakelandis should take a look here and give it a sign off?

joegallo replied:
That's a good suggestion -- I've hallway reviewed the code with him already, but I'll get an explicit +1 from him on this bit once we're doing with the rest.

jakelandis replied:
only reviewed the secure settings bits, LGTM pending the outcome of the error handling (see comment above)

dakrone reviewed (changes requested):
Review round two comments

masseyke commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportDeleteDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,132 @@
+
+            final String id = databaseId;
+
+            if (geoIpMeta.getDatabases().containsKey(id) == false) {
+                throw new ResourceNotFoundException("Database configuration not found: {}", id);
+            }
```
Since this check was already in the cluster state task, I just removed it from here.

joegallo replied:
We rewrote deletes a bit in https://github.com/elastic/elasticsearch/commit/2b5637115c414a5b68f0acf0ec62df499a62dbbe -- I marked the other comment about deletes as being resolved but I'm leaving this one unresolved as a marker for Lee to re-review the new code.

masseyke commented on x-pack/plugin/geoip-enterprise-downloader/src/main/java/org/elasticsearch/xpack/geoip/EnterpriseGeoIpDownloaderLicenseListener.java:
```diff
@@ -0,0 +1,145 @@
+
+    private void ensureTaskStarted() {
+        assert licenseIsValid : "Task should never be started without valid license";
+        persistentTasksService.sendStartRequest(
+            ENTERPRISE_GEOIP_DOWNLOADER,
+            ENTERPRISE_GEOIP_DOWNLOADER,
+            new EnterpriseGeoIpTaskParams(),
+            MASTER_TIMEOUT,
+            ActionListener.wrap(r -> logger.debug("Started enterprise geoip downloader task"), e -> {
+                Throwable t = e instanceof RemoteTransportException ? ExceptionsHelper.unwrapCause(e) : e;
+                if (t instanceof ResourceAlreadyExistsException == false) {
+                    logger.error("failed to create enterprise geoip downloader task", e);
+                }
+            })
+        );
```
We expect geoip metadata changes to be really rare (there are only 7 possible databases to download right now). And the task start request appears to be very lightweight. So it seemed a fair tradeoff, rather than a request to check if the task was running. Do you know of an inexpensive way to check if a task is running?

joegallo replied:
I talked through this one with @original-brownbear and he had some ideas on how we could write things to be better -- but he said this was fine enough as it was and that the better version (while better) is not urgently required to replace what's here.

masseyke committed:
Adding TransportPutDatabaseConfigurationActionTests

masseyke committed:
cleaning up and commenting EnterpriseGeoIpHttpFixture

masseyke commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpTaskState.java:
```diff
@@ -0,0 +1,153 @@
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.V_7_13_0;
```
Definitely.

masseyke committed:
fixing EnterpriseGeoIpTaskStates minimalSupportedVersion

masseyke committed:
avoiding keeping a collection of Map.Entry values in DatabaseNodeService

masseyke committed:
Cleanup of EnterpriseGeoIpDownloader -- adding comments, adding assertion messages, and incorporating code review feedback

masseyke committed:
renaming GeoIpTaskState.isValid() to isNewEnough()

joegallo committed:
Don't check for no-ops outside the cluster state thread

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/DatabaseConfigurationMetadata.java:
```diff
@@ -0,0 +1,83 @@
+            DatabaseConfiguration database = (DatabaseConfiguration) a[0];
+            return new DatabaseConfigurationMetadata(database, (long) a[1], (long) a[2]);
+        }
+    );
```
Agreed, f9dc8942ef3.

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/HttpClient.java:
```diff
@@ -34,6 +35,26 @@
 
 class HttpClient {
 
+    static final class PasswordAuthenticationHolder implements AutoCloseable {
```
77a76aebd74

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/IngestGeoIpMetadata.java:
```diff
@@ -0,0 +1,173 @@
+                DatabaseConfigurationMetadata databaseMeta = new DatabaseConfigurationMetadata(in);
+                databases.put(databaseMeta.database().id(), databaseMeta);
+            }
+            this.databases = Map.copyOf(databases);
```
Good catch, thanks! c48f91b2b21

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportGetDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,113 @@
+            }
+        } else {
+            listener.onResponse(new GetDatabaseConfigurationAction.Response(databases));
+        }
```
I changed the logic to be closer to what you asked via 8d43e18790c, but it's not precisely what you asked for -- feel free to re-comment on it if you don't like this third way.

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/IngestGeoIpMetadata.java:
```diff
@@ -0,0 +1,173 @@
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.MINIMUM_COMPATIBLE;
```
Yeah, we missed that one, thanks -- we did the right thing elsewhere and with 552eef73657 we're doing it here now, too.

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportDeleteDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,132 @@
+    protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<AcknowledgedResponse> listener)
+        throws Exception {
+        final String id = request.getDatabaseId();
+        deleteDatabaseConfigurationTaskQueue.submitTask(
```
We rewrote deletes a bit in 2b5637115c4 -- I'm going to mark this comment resolved, but I'm going to reference the same commit on the other comment and **not** resolve it because I would like you to review the new logic.

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportPutDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,198 @@
+            if (isNoopUpdate(existingDatabase, request.getDatabase())) {
+                listener.onResponse(AcknowledgedResponse.TRUE);
+                return;
+            }
```
I think bc7e4eb4e75 addresses this, but your concern is intricate enough that I'm leaving the resolve unresolved so that you can choose to resolve if you're happy with the new logic.

joegallo committed:
Remove my comments about this logic being mistaken

It's still true that I do think the code is mistaken, but I'll move
that notion into a github issue and link it for followup.

joegallo committed:
Switch to Tuples here, too

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseNodeService.java:
```diff
@@ -269,20 +271,42 @@ void checkDatabases(ClusterState state) {
-            // Note: an empty state will purge stale entries in databases map
-            taskState = GeoIpTaskState.EMPTY;
+        // we'll consult each of the geoip downloaders to build up a list of database metadatas to work with
+        List<Map.Entry<String, GeoIpTaskState.Metadata>> validMetadatas = new ArrayList<>();
```
Yeah, they're really only being used as a Tuple anyway, so that's fair -- fixed in e8b4f36409a.

joegallo committed:
Only set the flag if we actually added something

That is, don't consider a skip-due-to-credentials to be 'added
something'.

joegallo committed:
Drop the copyOf and simplify this slightly

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportGetDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,124 @@
+        final ActionListener<GetDatabaseConfigurationAction.Response> listener
+    ) {
+        IngestGeoIpMetadata geoIpMeta = state.metadata().custom(IngestGeoIpMetadata.TYPE, IngestGeoIpMetadata.EMPTY);
+        if (geoIpMeta.getDatabases().isEmpty()) {
+            if (request.getDatabaseIds().length == 0) {
+                // you asked for all, and there are none, we return the none that there are
+                listener.onResponse(new GetDatabaseConfigurationAction.Response(List.of()));
+            } else {
+                // you asked for *something*, and there are none, we 404
+                listener.onFailure(
+                    new ResourceNotFoundException(
+                        "database configuration or configurations {} not found, no database configurations are configured",
+                        Arrays.toString(request.getDatabaseIds())
+                    )
+                );
+            }
+        }
```
I think you accidentally removed the `return` statement here. In either of these cases we should return immediately since you've already invoked the `listener`, so the rest of the method isn't executed.

joegallo replied:
YUP. I fall upon my sword.

joegallo replied:
Fixed in 4fc4f46c3c017605c11a4e4dd9cd64bc5db44327.

joegallo committed:
Add back two missing early returns

dakrone commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/direct/TransportGetDatabaseConfigurationAction.java:
```diff
@@ -0,0 +1,124 @@
+                // you asked for all, and there are none, we return the none that there are
+                listener.onResponse(new GetDatabaseConfigurationAction.Response(List.of()));
+            } else {
+                // you asked for *something*, and there are none, we 404
+                listener.onFailure(
+                    new ResourceNotFoundException(
+                        "database configuration or configurations {} not found, no database configurations are configured",
+                        Arrays.toString(request.getDatabaseIds())
+                    )
+                );
```
One thing I'm still not sure that I like about this is that it treats `GET /_ingest/geoip/database/` and `GET /_ingest/geoip/database/*` differently. Are we sure we like that? I'm not sure I care for the divergent behavior.

joegallo replied:
I'm not married to it -- we have that behavior for SLM, so it's not without precedent, but it's not a detail that I'm super attached to. I'll simplify it out.

joegallo replied:
How does 90a749ca4077e6ee68311b0f09ee7c1f3748743c strike you?

dakrone replied:
I think that looks good to me.

joegallo committed:
Expand javadoc slightly

joegallo committed:
Don't handle the empty state case specially

joegallo committed:
No treble

joegallo committed:
Extend the fixture to cover an edge case

joegallo committed:
Merge branch 'main' into enterprise-downloader

joegallo committed:
Switch exception types and messages

joegallo committed:
Rename inner variable

joegallo committed:
Use a record instead of a Tuple

joegallo committed:
Wrap and rethrow rather than logging

joegallo committed:
Merge branch 'main' into enterprise-downloader

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloaderTaskExecutor.java:
```diff
@@ -0,0 +1,248 @@
+
+            @Override
+            public InputStream getFile(String setting) {
+                throw new IllegalStateException("A NotificationService setting cannot be File.");
```
Agreed, 572991a3822979d86eff7a41d38e10e73a8841e4

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloaderTaskExecutor.java:
```diff
@@ -0,0 +1,248 @@
+        try {
+            this.cachedSecureSettings = extractSecureSettings(settings, List.of(MAXMIND_LICENSE_KEY_SETTING));
+        } catch (GeneralSecurityException e) {
+            logger.error("Keystore exception while reloading enterprise geoip download task executor", e);
```
That makes sense, yeah. 95a820b9101d1680b4f8cb09517148ce5c5cf91c

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloaderTaskExecutor.java:
```diff
@@ -0,0 +1,248 @@
+        // get the secure settings out
+        final SecureSettings sourceSecureSettings = Settings.builder().put(source, true).getSecureSettings();
+        // filter and cache them...
+        final Map<String, Tuple<SecureString, byte[]>> cache = new HashMap<>();
```
I am in the business of saying yes to things: 12655ede31b2c3646e86f02c6d6276cb0e5e762d  and 31af5d6c7063c1da57f4a405a129624f69a8ded3

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

dakrone reviewed (approved):
LGTM, thanks for working through this with me

jakelandis reviewed (commented):
SecureSetting bits LGTM
